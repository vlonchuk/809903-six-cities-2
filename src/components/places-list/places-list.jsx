import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import Operation from './../../reducer/operation/operation.js';
import PlaceCard from './../place-card/place-card.jsx';
import PlacesListType from './../../consts/places-list-type.js';
import {showError} from './../../utils.js';
import Errors from './../../consts/errors.js';

class PlacesList extends PureComponent {
  constructor(props) {
    super(props);
    switch (this.props.listType) {
      case PlacesListType.CITY:
        this._placesClass = `cities__places-list tabs__content places__list`;
        break;
      case PlacesListType.NEAR:
        this._placesClass = `near-places__list`;
        break;
      case PlacesListType.FAVORITES:
        this._placesClass = `favorites__places`;
        break;
      default:
        this._placesClass = ``;
    }
  }

  render() {
    return <div className={`${this._placesClass}`}>
      {this.properties}
    </div>;
  }

  get properties() {
    return this.props.properties.map((item) => (
      <PlaceCard listType={this.props.listType} key={item.id} data={item} onClick={this.props.onClick}
        onMouseEnter={this.props.onPlaceCardMouseEnter}
        onMouseLeave={this.props.onPlaceCardMouseLeave}
        onAddToFavorite={this.props.onAddToFavorite}
      />
    ));
  }
}

PlacesList.propTypes = {
  listType: PropTypes.oneOf(Object.values(PlacesListType)).isRequired,
  properties: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    city: PropTypes.shape({
      name: PropTypes.string.isRequired,
      location: PropTypes.shape({
        latitude: PropTypes.number.isRequired,
        longitude: PropTypes.number.isRequired,
        zoom: PropTypes.number.isRequired,
      }),
    }),
    title: PropTypes.string.isRequired,
    previewImage: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    isPremium: PropTypes.bool.isRequired,
    rating: PropTypes.number.isRequired,
    location: PropTypes.shape({
      latitude: PropTypes.number.isRequired,
      longitude: PropTypes.number.isRequired,
      zoom: PropTypes.number.isRequired,
    }),
  })).isRequired,
  onClick: PropTypes.func,
  onPlaceCardMouseEnter: PropTypes.func,
  onPlaceCardMouseLeave: PropTypes.func,
  onAddToFavorite: PropTypes.func.isRequired,
};

const mapStateToProps = (_, ownProps) => ownProps;

const mapDispatchToProps = (dispatch) => ({
  onAddToFavorite: (hotelId, status, favorites) => {
    dispatch(Operation.addToFavorite(hotelId, status, favorites))
      .catch((err) => showError(err, Errors.ERR_ADD_FAVORITE));
  },
});

const PlacesListWrapped = connect(mapStateToProps, mapDispatchToProps)(PlacesList);
export {PlacesList};
export default PlacesListWrapped;
