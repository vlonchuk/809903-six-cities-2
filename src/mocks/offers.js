const offers = [
  {
    id: `prop-1`,
    city: `Amsterdam`,
    caption: `Beautiful & luxurious apartment at great location`,
    imgSrc: `img/apartment-01.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 120,
    priceText: `night`,
    rating: 50,
    coor: {
      latitude: 52.3909553943508,
      longitude: 4.85309666406198
    },
  },
  {
    id: `prop-2`,
    city: `Amsterdam`,
    caption: `Wood and stone place`,
    imgSrc: `img/room.jpg`,
    type: `Private room`,
    priceCurrency: `€`,
    priceValue: 80,
    priceText: `night`,
    rating: 80,
    coor: {
      latitude: 52.369553943508,
      longitude: 4.85309666406198
    },
  },
  {
    id: `prop-3`,
    city: `Amsterdam`,
    caption: `Canal View Prinsengracht`,
    imgSrc: `img/apartment-02.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 132,
    priceText: `night`,
    rating: 100,
    coor: {
      latitude: 52.3909553943508,
      longitude: 4.929309666406198
    },
  },
  {
    id: `prop-4`,
    city: `Amsterdam`,
    caption: `Nice, cozy, warm big bed apartment`,
    imgSrc: `img/apartment-03.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 180,
    priceText: `night`,
    rating: 65,
    coor: {
      latitude: 52.3809553943508,
      longitude: 4.939309666406198
    },
  },
  {
    id: `prop-5`,
    city: `Paris`,
    caption: `Canal View Prinsengracht`,
    imgSrc: `img/apartment-02.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 132,
    priceText: `night`,
    rating: 90,
    coor: {
      latitude: 52.3909553943508,
      longitude: 4.929309666406198
    },
  },
  {
    id: `prop-6`,
    city: `Paris`,
    caption: `Nice, cozy, warm big bed apartment`,
    imgSrc: `img/apartment-03.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 180,
    priceText: `night`,
    rating: 40,
    coor: {
      latitude: 52.3809553943508,
      longitude: 4.939309666406198
    },
  },
];

export const getPropertiesByCity = (city) => {
  return offers.filter((el) => el.city === city)
               .slice();
};

export default offers;
