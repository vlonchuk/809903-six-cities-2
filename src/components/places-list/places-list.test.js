import React from 'react';
import renderer from 'react-test-renderer';
import {PlacesList} from './places-list';
import offers from './../../mocks/offers.js';
import PlacesListType from './../../consts/places-list-type';

jest.mock(`./../place-card/place-card.jsx`, () => jest.fn().mockReturnValue(null));

it(`PlacesList correctly renders after relaunch`, () => {
  const tree = renderer
    .create(<PlacesList key="PlacesList" properties={offers}
      onPlaceCardMouseEnter={jest.fn()}
      onPlaceCardMouseLeave={jest.fn()}
      onAddToFavorite={jest.fn()}
      listType={PlacesListType.CITY}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
