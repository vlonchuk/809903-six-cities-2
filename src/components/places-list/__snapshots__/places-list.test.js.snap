// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesList correctly renders after relaunch 1`] = `
<div
  className="cities__places-list tabs__content places__list"
/>
`;
